File: c_test/matrix_mul.c
#include <stdio.h>
#define max_m_n 25

int main()
{
    int row_a = 0, row_b = 0, column_a = 0, column_b = 0; // 矩阵A/B的行列数
    int a[max_m_n][max_m_n], b[max_m_n][max_m_n];

    scanf("%d %d", &row_a, &column_a);
    for (int i = 0; i < row_a; i++)
        for (int j = 0; j < column_a; j++)
            scanf("%6d", &a[i][j]); // 输入A的矩阵
    scanf("%d %d", &row_b, &column_b);
    for (int i = 0; i < row_b; i++)
        for (int j = 0; j < column_b; j++)
            scanf("%6d", &b[i][j]); // 输入B的矩阵

    // 判断a和b是否可以执行矩阵相乘运算
    if (column_a == row_b ||
        (column_a == 1 && column_a == 1) ||
        (column_b == 1 && row_b == 1))
    { // 可以执行c=a*b运算
        int c[max_m_n][max_m_n];
        for (int i = 0; i < row_a; i++)
            for (int j = 0; j < column_b; j++)
            {
                c[i][j] = 0;
                for (int k = 0; k < row_b; k++)
                    c[i][j] += a[i][k] * b[k][j];
            }

        // 对c进行按格式输出
        for (int i = 0; i < row_a; i++)
        {
            for (int j = 0; j < column_b; j++)
                printf("%10d", c[i][j]);
            printf("\n"); // 换行
        }
    }
    else
    { // 无法进行矩阵乘法
        printf("Incompatible Dimensions\n");
    }
    return 0;
}

/*
例子 1
2 2
     1     2
     3     4
3 1
     1
     2
     3

例子 2
2 2
     1     2
     3     4
2 3
     1     2     0
     0     3     4

例子 3
4 1
   123
   345
   567
   789
1 4
  -123   345  -567   789
 */

File: pascal_test/matrixMul.pas
program matrixMul;
var
    a, b, result : array [0 .. 125] of longint; // 矩阵变量
    rowA, rowB, columnA, columnB: longint; // 矩阵A/B的行列数
    i, j, k, idx1, idx2, idx3 : integer;
begin
    read(rowA, columnA); // 输入A的矩阵
    for i := 0 to rowA - 1 do
    begin
        for j := 0 to columnA - 1 do
        begin
            idx1 := i * columnA + j;
            read(a[idx1]);
        end;
    end;

    read(rowB, columnB); // 输入B的矩阵
    for i := 0 to rowB - 1 do
    begin
        for j := 0 to columnB - 1 do
        begin
            idx1 := i * columnB + j;
            read(b[idx1]);
        end;
    end;

    // 判断a和b是否可以执行矩阵相乘运算
    if (columnA = rowB) or ((columnA = 1) and (rowA = 1)) or ((columnB = 1) and (rowB = 1)) then
    begin
        // 可以执行c=a*b运算
        for i := 0 to rowA - 1 do
        begin
            for j := 0 to columnB - 1 do
            begin
                idx1 := i * columnB + j;
                result[idx1] := 0;
                for k := 0 to rowB - 1 do
                begin
                    idx2 := i * columnA + k;
                    idx3 := k * columnB + j;
                    result[idx1] := result[idx1] + (a[idx2] * b[idx3]);
                end;
            end;
        end;

        // 对c进行按格式输出
        for i := 0 to rowA - 1 do
        begin
            for j := 0 to columnB - 1 do
            begin
                idx1 := i * columnB + j;
                write(result[idx1]: 10);
            end;
            writeln(); // 换行
        end;
    end

    else
    begin
        // 无法进行矩阵乘法
        writeln('Incompatible Dimensions')
    end;
end.